### ISSUE TITLE
Automate estimation of backlash

### ISSUE THREAD

**Issue by bgottula** [bgottula/track#7]
Estimation of backlash should be straightforward using optical tracking with a bright star. The motion of the star can be compared with the reported change in position of the mount. The result should be a pair of values in degrees for the azimuth and altitude directions.

**Comment by bgottula** [comment]
There are a couple of approaches I can think of:
1. Slew the mount back and forth along one axis (either azimuth or elevation) by some number of degrees per slew. Decrease the total angle of the slew on each iteration until the position of the object in the camera frame does not move. Then increase the total angle by very small increments on each slew until the object starts to move again. This could be implemented as a sort of binary search. At the beginning of the search the slew rate could be fast but near the end it should be slow to minimize error introduced by the lag between sending slew commands and reading back the mount position.
2. Slew the mount and intentionally move the target within the camera's field of view by a particular amount. Compare the motion of the object to the mount's reported change in position. The problem I see with this approach is that the accuracy will depend on the accuracy of conversion between motion of the object in the field of view (pixels vertical and horizontal) to degrees in azimuth and elevation. For the elevation axis this is straightforward, but at higher elevations the azimuth axis is not parallel to the horizontal axis of the frame. Getting this coordinate transformation right is not impossible but it would be more robust to eliminate this as a source of error.

Elaborating on the first approach, do the following for each axis. This assumes that a bright object (a star or planet) has been placed within the camera's FOV.
1. Center the object in the FOV (with software).
1. Slew on the selected axis in the + direction. Do not slew on the other axis.
1. Stop the slew when movement of the object is detected. Record the position of the object in the frame and the Az/Alt position of the mount.
1. Slew in the - direction until movement is detected. Record the positions.
1. Take the difference in the Az/Alt positions reported by the mount at the end of the two initial slews. This is the upper bound on the backlash. The backlash must be less than this because the object moved in the FOV on the second slew.
(more steps need to be filled in here)

On detecting position accurately: Near the start of the algorithm, it may be prudent to estimate the RMS level of noise in the object's detected position over a significant number of frames. This information could be used to determine how far the object needs to move for detection of movement to be reliable, and/or it could be used to determine how many frames should be averaged for each position estimate.

**Comment by bgottula** [comment]
The first round of testing revealed the following:

- The binary search was basically useless. The search would eventually find a value nearly equal to the value found on the initial search. It seems more useful to perform the initial search procedure (originally thought of as the upper bound search) several times and take the average.
- Default optical tracking loop parameters resulted in undamped oscillations, probably (and somewhat ironically) due to backlash. I had better results with a bandwidth of 0.1 Hz and a damping factor of 2.0.
- The default loop period of 0.25 s was too short. I had better luck with 0.5 s.
- Should add estimation of object velocity in FOV to do the following:
   - Subtract expected motion when detecting optical movement
   - Slew on the other axis to keep object in FOV
- May need to do something to keep object in the FOV during a single iteration. In the case I was testing (azimuth axis only) the star was moving mostly in the elevation axis due to sidereal motion. The star moved nearly off the top of the frame by the time the iteration was complete.

**Comment by bgottula** [comment]
The code now supports both azimuth and altitude axes. It now also continues to slew on the axis not under test at the same rate the tracking loop converged at in an effort to keep the object from drifting out of frame. I ran the script a few times this evening on a star but got wildly inconsistent results with the default parameters. I think the following refinements are necessary:
- As noted in the previous comment, estimate object's apparent sidereal motion and compare observed positions to predicted positions accounting for this motion.
- When detecting movement after the initial slew, this should be a comparison between the error before the slew is initiated rather than compared to zero error. Comparing with zero error assumes that the initial centering of the object was perfect. The default parameters set a wider bound on the initial centering than on the movement threshold, which sometimes causes the algorithm to "detect movement" even though the optical tube has not actually moved.
- Perhaps run each set of iterations twice in each axis, once starting in the positive slew direction and again starting in the negative direction. The two sets should be consistent if the object's apparent motion is compensated properly.

**Comment by bgottula** [comment]
Output of first complete run after algorithm overhaul. Getting more consistent results now.
[backlash_estimation_output_2017_08_09.txt](https://github.com/bgottula/track/files/1213775/backlash_estimation_output_2017_08_09.txt)

**Comment by bgottula** [comment]
While performing alignment testing I am seeing evidence that the first run significantly overestimated the backlash. Need to take a closer look at the results and figure out what went wrong.

**Comment by bgottula** [comment]
Some specific problems:

- Estimation of tracking slew rates is inconsistent between runs--perhaps dwell for a fixed time interval rather than relying on the TrackUntilConverged stopping condition
- Camera-to-mount rate ratio is not close to 1 when object sidereal motion is small
- Estimation of sidereal motion can take forever when rate is very slow, consider using a fixed time-based observation interval instead

**Comment by bgottula** [comment]
The star-tracking approach to backlash estimation has proven to be complex and disappointing. I have discovered a new approach that appears to be much simpler and more reliable. Rather than tracking a star, which introduces complexities due to the sidereal motion of the sky, I can track a terrestrial object that has no relative motion. I had avoided this previously because the current computer vision algorithm in the OpticalErrorSource class can't handle anything but a bright small object against a dark background. Fortunately this can be worked around by replacing computer vision with human vision. To aid in the detection of motion in the frame I found a bit of sample code that uses OpenCV to track the optical flow of features. Along with this visual aid I wrote a simple Python script that simply prints out the mount's change in position continually in both axes. By slewing the mount with the hand controller and watching the paths of the features in the OpenCV window it is reasonably straightforward to get a reliable estimate of the backlash deadbands. This new approach is not fully automated but since this is an activity that should not be required on a recurring basis I think the manual component is justified. In fact, seeing the motion in the camera frame provides much more confidence in the results. I suspect it would be too much trouble to achieve the same level of reliability with a fully automated solution.
